File: src/main/java/com/example/tutorial/common/utils/SampleDataLoader.java
package com.example.tutorial.common.utils;

import com.example.tutorial.common.datamodel.Meeting;
import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.datamodel.Task;
import com.example.tutorial.common.datamodel.Vehicle;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.function.Supplier;

public final class SampleDataLoader {

  private static final ObjectMapper MAPPER = new ObjectMapper()
      .registerModule(new JavaTimeModule())
      .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

  /**
   * Sample credit card offers data.
   */
  public static final Supplier<List<Offer>> OFFERS_DTO =
      () -> {
        try {
          String json = Files.readString(Paths.get("src/main/resources/sample_data/offer.json"));
          return (List<Offer>) MAPPER.readValue(json, new TypeReference<List<Offer>>() {
          });
        } catch (Exception e) {
          e.printStackTrace();
        }
        return List.of();
      };

  /**
   * Sample meeting schedule data for scheduling algorithms.
   * Use case: Meeting room allocation, calendar scheduling.
   */
  public static final Supplier<List<Meeting>> MEETINGS_DTO = () -> Arrays.asList(
      Meeting.builder().meetingId("MTG-001").title("Sprint Planning").startTime(540).endTime(600)
          .organizer("John").build(),    // 9:00-10:00
      Meeting.builder().meetingId("MTG-002").title("Design Review").startTime(570).endTime(630)
          .organizer("Sarah").build(),   // 9:30-10:30 (overlaps with MTG-001)
      Meeting.builder().meetingId("MTG-003").title("Daily Standup").startTime(600).endTime(615)
          .organizer("Team").build(),    // 10:00-10:15
      Meeting.builder().meetingId("MTG-004").title("Client Call").startTime(660).endTime(720)
          .organizer("Mike").build(),    // 11:00-12:00
      Meeting.builder().meetingId("MTG-005").title("Lunch & Learn").startTime(720).endTime(780)
          .organizer("HR").build(),      // 12:00-13:00
      Meeting.builder().meetingId("MTG-006").title("Code Review").startTime(840).endTime(900)
          .organizer("Dev Team").build(),// 14:00-15:00
      Meeting.builder().meetingId("MTG-007").title("1:1 with Manager").startTime(900).endTime(930)
          .organizer("Jane").build(),    // 15:00-15:30
      Meeting.builder().meetingId("MTG-008").title("Architecture Discussion").startTime(870).endTime(960)
          .organizer("Tech Lead").build() // 14:30-16:00 (overlaps with MTG-006, MTG-007)
  );

  /**
   * Sample vehicle data for parking lot algorithms.
   * Use case: Parking lot management, resource allocation.
   */
  public static final Supplier<List<Vehicle>> VEHICLES_DTO = () -> Arrays.asList(
      Vehicle.builder().vehicleId("VEH-001").licensePlate("ABC-1234").type(Vehicle.VehicleType.SMALL)
          .owner("Alice").build(),
      Vehicle.builder().vehicleId("VEH-002").licensePlate("XYZ-5678").type(Vehicle.VehicleType.MEDIUM)
          .owner("Bob").build(),
      Vehicle.builder().vehicleId("VEH-003").licensePlate("DEF-9012").type(Vehicle.VehicleType.BIG)
          .owner("Charlie").build(),
      Vehicle.builder().vehicleId("VEH-004").licensePlate("GHI-3456").type(Vehicle.VehicleType.SMALL)
          .owner("Diana").build(),
      Vehicle.builder().vehicleId("VEH-005").licensePlate("JKL-7890").type(Vehicle.VehicleType.MEDIUM)
          .owner("Eve").build(),
      Vehicle.builder().vehicleId("VEH-006").licensePlate("MNO-2345").type(Vehicle.VehicleType.SMALL)
          .owner("Frank").build()
  );

  /**
   * Sample task data for producer-consumer and concurrent processing.
   * Use case: Task queues, thread pool demos, async processing.
   */
  public static final Supplier<List<Task>> TASKS_DTO = () -> Arrays.asList(
      Task.builder().taskId("TASK-001").name("Process Payment").priority(Task.Priority.HIGH)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-002").name("Send Notification").priority(Task.Priority.MEDIUM)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-003").name("Generate Report").priority(Task.Priority.LOW)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-004").name("Validate Transaction").priority(Task.Priority.CRITICAL)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-005").name("Update Cache").priority(Task.Priority.MEDIUM)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-006").name("Sync Database").priority(Task.Priority.HIGH)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-007").name("Archive Logs").priority(Task.Priority.LOW)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build(),
      Task.builder().taskId("TASK-008").name("Fraud Check").priority(Task.Priority.CRITICAL)
          .status(Task.Status.PENDING).createdAt(System.currentTimeMillis()).build()
  );
}

File: src/main/python/com/example/tutorial/common/utils/sample_data_loader.py
"""
Sample Data Loader
----------------------------------
This module loads sample data from JSON files and provides factory methods.
Equivalent to the Java SampleDataLoader.java class.
"""

import json
import os
from typing import List
from com.example.tutorial.common.datamodel.offer import Offer, Partner, Merchant, Eligibility
from com.example.tutorial.common.datamodel.meeting import Meeting
from com.example.tutorial.common.datamodel.vehicle import Vehicle, VehicleType
from com.example.tutorial.common.datamodel.task import Task, Priority


def _parse_offer(data: dict) -> Offer:
    """Parse a single Offer from JSON dict."""
    partner_data = data.get("partner", {})
    merchant_data = data.get("merchant", {})
    eligibility_data = data.get("eligibility", {})

    return Offer(
        offer_id=data.get("offerId"),
        campaign_id=data.get("campaignId"),
        partner=Partner(
            partner_id=partner_data.get("partnerId"),
            name=partner_data.get("name")
        ) if partner_data else None,
        merchant=Merchant(
            merchant_id=merchant_data.get("merchantId"),
            name=merchant_data.get("name"),
            mcc=merchant_data.get("mcc"),
            brands=merchant_data.get("brands", [])
        ) if merchant_data else None,
        title=data.get("title"),
        description=data.get("description"),
        eligibility=Eligibility(
            categories=eligibility_data.get("categories", [])
        ) if eligibility_data else None
    )


def load_offers() -> List[Offer]:
    """
    Load offers from the sample JSON file.

    Returns:
        List of Offer objects parsed from offer.json
    """
    json_path = os.path.join(
        os.path.dirname(__file__), "..", "..", "..", "..",
        "resources", "sample_data", "offer.json"
    )
    json_path = os.path.normpath(json_path)

    try:
        with open(json_path, 'r') as f:
            data = json.load(f)
            return [_parse_offer(offer_data) for offer_data in data]
    except Exception as e:
        print(f"Error loading offers: {e}")
        return []


def load_meetings() -> List[Meeting]:
    """
    Load sample meeting data for scheduling algorithms.

    Returns:
        List of Meeting objects
    """
    return [
        Meeting.of_hours("Team Standup", 9, 10, "Alice"),
        Meeting.of_hours("Sprint Planning", 10, 12, "Bob"),
        Meeting.of_hours("Design Review", 11, 13, "Charlie"),
        Meeting.of_hours("1:1 Meeting", 14, 15, "Alice"),
        Meeting.of_hours("Tech Talk", 15, 17, "Diana"),
    ]


def load_vehicles() -> List[Vehicle]:
    """
    Load sample vehicle data for parking algorithms.

    Returns:
        List of Vehicle objects
    """
    return [
        Vehicle.of("ABC-1234", VehicleType.SMALL),
        Vehicle.of("XYZ-5678", VehicleType.MEDIUM),
        Vehicle.of("DEF-9012", VehicleType.BIG),
        Vehicle.of("GHI-3456", VehicleType.SMALL),
        Vehicle.of("JKL-7890", VehicleType.MEDIUM),
    ]


def load_tasks() -> List[Task]:
    """
    Load sample task data for producer-consumer patterns.

    Returns:
        List of Task objects
    """
    return [
        Task.of("Process Payment", Priority.HIGH),
        Task.of("Send Notification", Priority.MEDIUM),
        Task.of("Generate Report", Priority.LOW),
        Task.of("Update Cache", Priority.CRITICAL),
        Task.of("Sync Database", Priority.HIGH),
        Task.of("Archive Logs", Priority.LOW),
    ]
